graph/src/blockchain/block_stream.rs

```diff
@@ -316,19 +315,155 @@ impl<C: Blockchain> BlockWithTriggers<C> {
 
 pub struct TriggersAdapterWrapper<C: Blockchain> {
     pub adapter: Arc<dyn TriggersAdapter<C>>,
-    pub source_subgraph_stores: Vec<(DeploymentHash, Arc<dyn WritableStore>)>,
+    pub source_subgraph_stores: HashMap<DeploymentHash, Arc<dyn WritableStore>>,
 }
 
 impl<C: Blockchain> TriggersAdapterWrapper<C> {
     pub fn new(
         adapter: Arc<dyn TriggersAdapter<C>>,
-        source_subgraph_stores: Vec<(DeploymentHash, Arc<dyn WritableStore>)>,
+        source_subgraph_stores: HashMap<DeploymentHash, Arc<dyn WritableStore>>,
     ) -> Self {
         Self {
             adapter,
             source_subgraph_stores,
         }
     }
+
+    pub async fn blocks_with_subgraph_triggers(
+        &self,
+        logger: &Logger,
+        subgraph_filter: &SubgraphFilter,
+        range: SubgraphTriggerScanRange<C>,
+    ) -> Result<Vec<BlockWithTriggers<C>>, Error> {
+        let store = self
+            .source_subgraph_stores
+            .get(&subgraph_filter.subgraph)
+            .unwrap(); // TODO(krishna): Avoid unwrap
+
+        let schema = crate::components::store::ReadStore::input_schema(store);
+
+        let adapter = self.adapter.clone();
+
+        scan_subgraph_triggers::<C>(logger, store, &adapter, &schema, &subgraph_filter, range).await
+    }
+}
+
+fn create_subgraph_trigger_from_entities(
+    filter: &SubgraphFilter,
+    entities: &Vec<EntityWithType>,
+) -> Vec<subgraph::TriggerData> {
+    entities
+        .iter()
+        .map(|e| subgraph::TriggerData {
+            source: filter.subgraph.clone(),
+            entity: e.entity.clone(),
+            entity_type: e.entity_type.as_str().to_string(),
+        })
+        .collect()
+}
+
+async fn create_subgraph_triggers<C: Blockchain>(
+    logger: Logger,
+    blocks: Vec<C::Block>,
+    filter: &SubgraphFilter,
+    entities: BTreeMap<BlockNumber, Vec<EntityWithType>>,
+) -> Result<Vec<BlockWithTriggers<C>>, Error> {
+    let logger_clone = logger.cheap_clone();
+
+    let blocks: Vec<BlockWithTriggers<C>> = blocks
+        .into_iter()
+        .map(|block| {
+            let block_number = block.number();
+            match entities.get(&block_number) {
+                Some(e) => {
+                    let trigger_data = create_subgraph_trigger_from_entities(filter, e);
+                    BlockWithTriggers::new_with_subgraph_triggers(
+                        block,
+                        trigger_data,
+                        &logger_clone,
+                    )
+                }
+                None => BlockWithTriggers::new_with_subgraph_triggers(block, vec![], &logger_clone),
+            }
+        })
+        .collect();
+
+    Ok(blocks)
+}
+
+pub enum SubgraphTriggerScanRange<C: Blockchain> {
+    Single(C::Block),
+    Range(BlockNumber, BlockNumber),
+}
+
+async fn scan_subgraph_triggers<C: Blockchain>(
```

Comment:
is there any testing on these, specifically to ensure the comments on the code hold true like the range is inclusive etc

Reply:
I will dedicate a PR to more intensive testing in the end of this PR chain so that i have all the relevant changes ready before the testing